cgroup1/utils.go

```diff
@@ -130,12 +132,23 @@ func hugePageSizes() ([]string, error) {
 	return pageSizes, nil
 }
 
-func readUint(path string) (uint64, error) {
-	v, err := os.ReadFile(path)
+// readUint reads a uint64 from the given file and uses the given buffer as
+// scratch space. The buffer is intended to be reused across calls.
+func readUint(path string, buf *bytes.Buffer) (uint64, error) {
```

Comment:
Is this threadsafe anymore though? If you called any of the methods that now invoke this in a couple goroutines, there's no guarantee we'll get back the value we're expecting as another read might finish before, you may end up with interleaved data as shown below, or another thread might hit the `buf.Reset` before another gets to calling `buf.String()` and you'd end up trying to parse an empty string.

```go
func readUintHelper(path string, expected uint64, buf *bytes.Buffer) error {
	actual, err := readUint(path, buf)
	if err != nil {
		return err
	}
	if expected != actual {
		return fmt.Errorf("expected %d, got %d", expected, actual)
	}
	return nil
}

func TestReadUint(t *testing.T) {
	var (
		buf = bytes.NewBuffer(nil)
		eg  errgroup.Group
	)
	eg.Go(func() error {
		return readUintHelper("testdata/uint100", 100, buf)
	})
	eg.Go(func() error {
		return readUintHelper("testdata/uint200", 200, buf)
	})
	eg.Go(func() error {
		return readUintHelper("testdata/uint300", 300, buf)
	})

	if err := eg.Wait(); err != nil {
		t.Fatal(err)
	}
}
```

Yields:
```shell
dcantah@dcantah ~/go/src/github.com/containerd/cgroups ((HEAD detached at brancz/readuint)) $ go test ./cgroup1/ -v -run TestReadUint
=== RUN   TestReadUint
    utils_test.go:66: expected 100, got 300100
--- FAIL: TestReadUint (0.00s)
FAIL
FAIL    github.com/containerd/cgroups/v3/cgroup1        0.003s
FAIL
dcantah@dcantah ~/go/src/github.com/containerd/cgroups ((HEAD detached at brancz/readuint)) $ go test ./cgroup1/ -v -run TestReadUint
=== RUN   TestReadUint
    utils_test.go:66: expected 100, got 200100
--- FAIL: TestReadUint (0.00s)
FAIL
FAIL    github.com/containerd/cgroups/v3/cgroup1        0.002s
FAIL
dcantah@dcantah ~/go/src/github.com/containerd/cgroups ((HEAD detached at brancz/readuint)) $ go test ./cgroup1/ -v -run TestReadUint
=== RUN   TestReadUint
    utils_test.go:66: strconv.ParseUint: parsing "": invalid syntax
--- FAIL: TestReadUint (0.00s)
FAIL
FAIL    github.com/containerd/cgroups/v3/cgroup1        0.002s
FAIL
```

Reply:
Oh nice, I forgot the cg1 code has a lock around most of these for the main *cgroup operations. Let me see if there's any that the above might be a concern

Reply:
At the very least if you created individual controllers (e.g. cgroup1.NewMemory() -> mem.Stat()) you'd run into this

Reply:
Yeah I was briefly contemplating using a `sync.Pool`, but the way I understood the flow of the code there is no concurrency in the controllers. It’s entirely possible I missed the code path that does, so if that’s the case I’d happily change it to a pool to make it concurrency safe.

Reply:
In the controllers themselves there's not, but the library doesn't dissuade you from doing this yourself. How containerd uses this library is mainly through manipulating the main `Cgroup` interface from `New` or `Load` (https://github.com/containerd/cgroups/blob/main/cgroup1/cgroup.go#L37), and Stat and a bunch of others for this object [lock as they collect the data from each controller/subsystem](https://github.com/containerd/cgroups/blob/main/cgroup1/cgroup.go#L299-L326), so containerd's usage may be fine. 

The library offers you the ability to Create individual controllers however with New{Controller} functions ([NewMemory](https://github.com/containerd/cgroups/blob/main/cgroup1/memory.go#L157) for example). These don't have any locking around them, so if anyone using this library relied on making an individual controller and then spinning up some goroutines to grab/update data, they may run into this now.

Reply:
You’re absolutely right. How do you propose to move forward? I would say we can either analyze usages of the library and if there are no current concurrent usages we add a code comment saying they are not concurrency safe, or we go straight to the hammer and use a buffer pool?

I generally prefer to not add synchronization if I don’t absolutely have to, but ultimately your call as a maintainer.

Reply:
You wanna benchmark the buffer pool idea? It'd be a shame if this caused us to not be able to freely call these however we want. Thanks for the quick follow ups! This is awesome to see btw, very cool stream idea

Reply:
I think the problem before was that `os.ReadFile` _returns_ a byte slice which forces an allocation because it escapes the stack. 

You can make this much simpler and still thread safe by creating a slice with `make([]byte, 64)` and `f.Read` to read the data. Since that slice isn't returned, it stays on the stack.

Reply:
Oh fantastic point duh, can confirm it shouldn't escape:

```go
package main

import (
	"log"
	"os"
)

func readFile() error {
	f, err := os.Open("howdy.txt")
	if err != nil {
		return err
	}
	defer f.Close()

	howdy := make([]byte, 64)
	_, err = f.Read(howdy)
	if err != nil {
		return err
	}
	return nil
}

func main() {
	if err := readFile(); err != nil {
		log.Fatal(err)
	}
}

```

```shell
alloc go build -gcflags '-m -l'
# github.com/dcantah/fun/alloc
./main.go:15:15: make([]byte, 64) does not escape
./main.go:25:12: ... argument does not escape
```

Reply:
@brancz Want to take this for a spin?